src/engine/SortedUnionImpl.h

```diff
@@ -0,0 +1,241 @@
+          origin == Union::NO_COLUMN ? Id::makeUndefined() : row[origin];
+    }
+  }
+
+  // Increment the iterators if the current id table is fully processed and
+  // reset the index back to zero.
+  void advanceRangeIfConsumed() {
+    if (index1_ == it1_.value()->idTable_.size()) {
+      ++it1_.value();
+      index1_ = 0;
+    }
+    if (index2_ == it2_.value()->idTable_.size()) {
+      ++it2_.value();
+      index2_ = 0;
+    }
+  }
+
+  // Retrieve the current result from `resultTable_` and `localVocab_` and reset
+  // those members back to their initial value so the next operation can
+  // continue adding values.
+  Result::IdTableVocabPair popResult() {
+    auto result = Result::IdTableVocabPair{std::move(resultTable_),
+                                           std::move(localVocab_)};
+    resultTable_ = IdTable{resultTable_.numColumns(), allocator_};
+    resultTable_.reserve(Union::chunkSize);
+    localVocab_ = LocalVocab{};
+    return result;
+  }
+
+  // ___________________________________________________________________________
+  std::optional<Result::IdTableVocabPair> get() override {
+    if (aggregatedTableReturned_) {
+      return std::nullopt;
+    }
+    if (!it1_.has_value()) {
+      it1_ = range1_.begin();
+    }
+    if (!it2_.has_value()) {
+      it2_ = range2_.begin();
+    }
+    while (it1_ != range1_.end() && it2_ != range2_.end()) {
+      auto& idTable1 = it1_.value()->idTable_;
+      auto& idTable2 = it2_.value()->idTable_;
+      localVocab_.mergeWith(std::span{&it1_.value()->localVocab_, 1});
+      localVocab_.mergeWith(std::span{&it2_.value()->localVocab_, 1});
+      while (index1_ < idTable1.size() && index2_ < idTable2.size()) {
+        if (isSmaller(idTable1.at(index1_), idTable2.at(index2_))) {
+          pushRow(true, idTable1.at(index1_));
+          index1_++;
+        } else {
+          pushRow(false, idTable2.at(index2_));
+          index2_++;
+        }
+        if (requestLaziness_ && resultTable_.size() >= Union::chunkSize) {
+          auto result = popResult();
+          advanceRangeIfConsumed();
+          return result;
+        }
+      }
+      advanceRangeIfConsumed();
+    }
+    if (requestLaziness_) {
+      if (index1_ != 0) {
+        appendCurrent(leftPermutation_, it1_.value(), index1_);
+      }
+      if (index2_ != 0) {
+        appendCurrent(rightPermutation_, it2_.value(), index2_);
+      }
+      if (!resultTable_.empty()) {
+        return Result::IdTableVocabPair{std::move(resultTable_),
+                                        std::move(localVocab_)};
+      }
+      auto leftTable =
+          passNext(leftPermutation_, it1_.value(), range1_.end(), index1_);
+      return leftTable.has_value() ? std::move(leftTable)
+                                   : passNext(rightPermutation_, it2_.value(),
+                                              range2_.end(), index2_);
+    }
+    appendRemaining(leftPermutation_, it1_.value(), range1_.end(), index1_);
```

Comment:
sprinkel in the redundant correctness check for fully materialized results as a documentation.

Reply:
The if statement before is literally
```cpp
if (requestLaziness_) {
  // ...
  return /*...*/;
}
```
So I don't think a correctness check makes sense here.